### PR TITLE
fix: MySQL tests

### PR DESCRIPTION
**Pull request by kenjis**
See https://github.com/codeigniter4/shield/pull/331#issuecomment-1202085326

- set property cast `integer` for `User::$id`

```console
$ composer test
> phpunit
PHPUnit 9.5.21 #StandWithUkraine

Runtime:       PHP 8.0.21
Configuration: /Users/kenji/work/codeigniter/codeigniter-shield/phpunit.xml
Random Seed:   1659515574

...............................................................  63 / 218 ( 28%)
............................................................... 126 / 218 ( 57%)
............................................................... 189 / 218 ( 86%)
.............................                                   218 / 218 (100%)

Nexus\PHPUnit\Extension\Tachycardia identified these 30 slow tests:
+----------------------------------------------------------------------------------------------------------------+---------------+-------------+
| Test Case                                                                                                      | Time Consumed | Time Limit  |
+----------------------------------------------------------------------------------------------------------------+---------------+-------------+
| Tests\\Controllers\\ActionsTest::testEmail2FAVerify                                                            | 00:00:02.45   | 00:00:00.50 |
| Tests\\Controllers\\RegisterTest::testRegisterActionSuccess                                                    | 00:00:01.95   | 00:00:00.50 |
| Tests\\Controllers\\RegisterTest::testRegisterRedirectsToActionIfDefined                                       | 00:00:01.89   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testEmailActivateVerify                                                       | 00:00:01.81   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testEmailActivateCannotBeBypassed                                             | 00:00:01.68   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testEmail2FAHandleSendsEmail                                                  | 00:00:01.45   | 00:00:00.50 |
| Tests\\Authentication\\Filters\\TokenFilterTest::testFilterSuccess                                             | 00:00:01.44   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testLoggedInFalse                             | 00:00:01.28   | 00:00:00.50 |
| Tests\\Authorization\\AuthorizableTest::testAddPermissionWithExistingPermissions                               | 00:00:01.24   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\AccessTokenAuthenticatorTest::testCheckBadToken                         | 00:00:01.12   | 00:00:00.50 |
| Tests\\Unit\\DictionaryValidatorTest::testCheckTrueOnNotFound                                                  | 00:00:01.09   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testEmail2FAVerifyFails                                                       | 00:00:01.06   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptCaseInsensitive                    | 00:00:01.05   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testEmail2FACannotBeBypassed                                                  | 00:00:01.05   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testShowEmail2FACreatesIdentity                                               | 00:00:01.05   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testEmail2FAShow                                                              | 00:00:01.04   | 00:00:00.50 |
| Tests\\Controllers\\MagicLinkTest::testAfterLoggedInNotAllowDisplayMagicLink                                   | 00:00:01.02   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testForgetAnotherUser                         | 00:00:01.01   | 00:00:00.50 |
| Tests\\Authentication\\Filters\\TokenFilterTest::testFilterNotAuthorized                                       | 00:00:01.00   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testLoginByIdRemember                         | 00:00:00.98   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testEmailActivateShow                                                         | 00:00:00.94   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testGetIdentitiesByType                                                                 | 00:00:00.94   | 00:00:00.50 |
| Tests\\Unit\\UserIdentityModelTest::testCreateDuplicateRecordThrowsException                                   | 00:00:00.94   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptUserHavingSessionDataAttemptsAgain | 00:00:00.93   | 00:00:00.50 |
| Tests\\Authorization\\AuthorizableTest::testAddPermissionsWithUnknownPermission                                | 00:00:00.92   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testLoginRedirectsToActionIfDefined                                             | 00:00:00.91   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testLoginWithRemember                         | 00:00:00.90   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptCannotFindUser                     | 00:00:00.90   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptUsernameOnly                       | 00:00:00.90   | 00:00:00.50 |
| Tests\\Unit\\UserIdentityModelTest::testCreateCodeIdentityThrowsExceptionIfUniqueCodeIsNotGot                  | 00:00:00.89   | 00:00:00.50 |
+----------------------------------------------------------------------------------------------------------------+---------------+-------------+
...and 101 more tests hidden from view.


Time: 01:59.134, Memory: 42.00 MB

OK (218 tests, 506 assertions)
```

### CODE REVIEW

File: src/Entities/User.php [hunk 1]
```diff
@@ -40,6 +40,7 @@ class User extends Entity
      * @var array<string, string>
      */
     protected $casts = [
+        'id'               => 'integer',
         'active'           => 'boolean',
         'force_pass_reset' => 'boolean',
```
Comment (kenjis) [inline src/Entities/User.php:45]: What's `force_pass_reset`?

Comment (MGatner) [reply src/Entities/User.php:45]: Probably just a paste issue from Myth: https://github.com/lonnieezell/myth-auth/blob/develop/src/Entities/User.php

File: src/Models/RememberModel.php [hunk 2]
```diff
@@ -23,6 +25,16 @@ class RememberModel extends Model
     ];
     protected $useTimestamps = true;
 
+    public function fake(Generator &$faker): stdClass
+    {
+        return (object) [
+            'user_id'         => fake(UserModel::class)->id,
```
[the remark this answers is not in the document]
Comment (kenjis) [reply src/Models/RememberModel.php]: Without thinking, I just copied the code from another model.

We don't need a user record. Fixed.

File: src/Entities/User.php [hunk 3]
```diff
@@ -40,6 +40,7 @@ class User extends Entity
      * @var array<string, string>
      */
     protected $casts = [
+        'id'               => 'integer',
```
[the remark this answers is not in the document]
Comment (kenjis) [reply src/Entities/User.php]: Oh, thanks! I forgot about nullable!
Fixed.

File: src/Models/UserModel.php [hunk 4]
```diff
@@ -298,7 +298,9 @@ protected function saveEmailIdentity(array $data): array
         }
 
         // Insert
-        if ($this->tempUser->id === null) {
+        // If we use Entity Property Casting ['id' => 'integer'],
+        // `id` returns 0 when not set.
```
[the remark this answers is not in the document]
Comment (kenjis) [reply src/Models/UserModel.php]: Done.

**Comment by kenjis** [comment]
```console
$ composer test
> phpunit
PHPUnit 9.5.21 #StandWithUkraine

Runtime:       PHP 8.0.21
Configuration: /Users/kenji/work/codeigniter/codeigniter-shield/phpunit.xml
Random Seed:   1659565388

...............................................................  63 / 218 ( 28%)
............................................................... 126 / 218 ( 57%)
............................................................... 189 / 218 ( 86%)
.............................                                   218 / 218 (100%)

Nexus\PHPUnit\Extension\Tachycardia identified these 30 slow tests:
+----------------------------------------------------------------------------------------------------------------+---------------+-------------+
| Test Case                                                                                                      | Time Consumed | Time Limit  |
+----------------------------------------------------------------------------------------------------------------+---------------+-------------+
| Tests\\Controllers\\RegisterTest::testRegisterRedirectsToActionIfDefined                                       | 00:00:01.61   | 00:00:00.50 |
| Tests\\Controllers\\RegisterTest::testRegisterActionSuccess                                                    | 00:00:01.60   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testModelfindAllWithIdentities                                                          | 00:00:01.10   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testModelfindByIdWithIdentities                                                         | 00:00:01.05   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testLoginRedirectsToActionIfDefined                                             | 00:00:01.00   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testUpdatePassword                                                                      | 00:00:00.97   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptUserHavingSessionDataAttemptsAgain | 00:00:00.97   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testLoginBadEmail                                                               | 00:00:00.95   | 00:00:00.50 |
| Tests\\Controllers\\RegisterTest::testRegisterActionRedirectsIfLoggedIn                                        | 00:00:00.95   | 00:00:00.50 |
| Tests\\Authentication\\HasAccessTokensTest::testRevokeAllAccessTokens                                          | 00:00:00.95   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptUsernameOnly                       | 00:00:00.91   | 00:00:00.50 |
| Tests\\Authorization\\AuthorizableTest::testAddGroupWithUnknownGroup                                           | 00:00:00.91   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testLastLogin                                                                           | 00:00:00.90   | 00:00:00.50 |
| Tests\\Controllers\\MagicLinkTest::testAfterLoggedInNotAllowDisplayMagicLink                                   | 00:00:00.90   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testLoginWithRemember                         | 00:00:00.88   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testLoginActionUsernameSuccess                                                  | 00:00:00.88   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testUpdatePasswordHash                                                                  | 00:00:00.87   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testEmailActivateShow                                                         | 00:00:00.86   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testAfterLoggedInNotDisplayLoginPage                                            | 00:00:00.86   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testGetIdentitiesSome                                                                   | 00:00:00.85   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testEmail2FAVerify                                                            | 00:00:00.85   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testPreviousLogin                                                                       | 00:00:00.85   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testGetIdentitiesByType                                                                 | 00:00:00.85   | 00:00:00.50 |
| Tests\\Authentication\\MagicLinkTest::testMagicLinkVerifyExpired                                               | 00:00:00.85   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testLoginActionEmailSuccess                                                     | 00:00:00.85   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testUpdateEmail                                                                         | 00:00:00.84   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptSuccess                            | 00:00:00.83   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptCaseInsensitive                    | 00:00:00.82   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testCheckBadPassword                          | 00:00:00.81   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testShowEmail2FACreatesIdentity                                               | 00:00:00.80   | 00:00:00.50 |
+----------------------------------------------------------------------------------------------------------------+---------------+-------------+
...and 106 more tests hidden from view.


Time: 01:48.687, Memory: 42.00 MB

OK (218 tests, 506 assertions)
```

File: src/Entities/User.php (hunk 1 above)
Comment (kenjis) [reply src/Entities/User.php:45]: Removed. #337

**Comment by kenjis** [comment]
```console
$ composer test
> phpunit
PHPUnit 9.5.21 #StandWithUkraine

Runtime:       PHP 8.0.21
Configuration: /Users/kenji/work/codeigniter/codeigniter-shield/phpunit.xml
Random Seed:   1659602347

...............................................................  63 / 225 ( 28%)
............................................................... 126 / 225 ( 56%)
............................................................... 189 / 225 ( 84%)
....................................                            225 / 225 (100%)

Nexus\PHPUnit\Extension\Tachycardia identified these 30 slow tests:
+--------------------------------------------------------------------------------------------+---------------+-------------+
| Test Case                                                                                  | Time Consumed | Time Limit  |
+--------------------------------------------------------------------------------------------+---------------+-------------+
| Tests\\Controllers\\RegisterTest::testRegisterActionSuccess                                | 00:00:02.40   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testLoginRedirectsToActionIfDefined                         | 00:00:02.29   | 00:00:00.50 |
| Tests\\Controllers\\RegisterTest::testRegisterRedirectsToActionIfDefined                   | 00:00:02.16   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testPreviousLogin                                                   | 00:00:02.11   | 00:00:00.50 |
| Tests\\Authorization\\AuthorizableTest::testCanCascadesToGroupsWithWildcards               | 00:00:01.64   | 00:00:00.50 |
| Tests\\Unit\\UserModelTest::testUpdateUserObjectWithoutUserDataToUpdate                    | 00:00:01.48   | 00:00:00.50 |
| Tests\\Authentication\\MagicLinkTest::testMagicLinkSubmitSuccess                           | 00:00:01.25   | 00:00:00.50 |
| Tests\\Unit\\UserModelTest::testSaveInsertUser                                             | 00:00:01.24   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testLogoutAction                                            | 00:00:01.21   | 00:00:00.50 |
| Tests\\Authentication\\Filters\\ChainFilterTest::testFilterSuccessSeession                 | 00:00:01.21   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testGetIdentitiesSome                                               | 00:00:01.20   | 00:00:00.50 |
| Tests\\Unit\\DictionaryValidatorTest::testCheckTrueOnNotFound                              | 00:00:01.16   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\AccessTokenAuthenticatorTest::testCheckOldToken     | 00:00:01.15   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testLoginActionEmailSuccess                                 | 00:00:01.09   | 00:00:00.50 |
| Tests\\Controllers\\RegisterTest::testRegisterDisplaysForm                                 | 00:00:01.08   | 00:00:00.50 |
| Tests\\Unit\\UserModelTest::testSaveUpdateUserObjectWithUserDataToUpdate                   | 00:00:01.05   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptCannotFindUser | 00:00:01.05   | 00:00:00.50 |
| Tests\\Controllers\\LoginTest::testLoginActionUsernameSuccess                              | 00:00:01.04   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testUpdatePasswordHash                                              | 00:00:01.04   | 00:00:00.50 |
| Tests\\Authentication\\Authenticators\\SessionAuthenticatorTest::testAttemptUsernameOnly   | 00:00:01.03   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testModelfindAllWithIdentities                                      | 00:00:01.03   | 00:00:00.50 |
| Tests\\Controllers\\MagicLinkTest::testAfterLoggedInNotAllowDisplayMagicLink               | 00:00:01.00   | 00:00:00.50 |
| Tests\\Authorization\\AuthorizableTest::testAddPermissionWithNoExistingPermissions         | 00:00:01.00   | 00:00:00.50 |
| Tests\\Controllers\\ActionsTest::testActionShowNoneAvailable                               | 00:00:00.99   | 00:00:00.50 |
| Tests\\Unit\\UserModelTest::testUpdateUserArrayWithUserDataToUpdate                        | 00:00:00.98   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testModelfindByIdWithIdentities                                     | 00:00:00.98   | 00:00:00.50 |
| Tests\\Unit\\UserTest::testGetIdentitiesByType                                             | 00:00:00.95   | 00:00:00.50 |
| Tests\\Authentication\\Filters\\TokenFilterTest::testFilterSuccess                         | 00:00:00.95   | 00:00:00.50 |
| Tests\\Authentication\\MagicLinkTest::testMagicLinkSubmitNoEmail                           | 00:00:00.94   | 00:00:00.50 |
| Tests\\Authorization\\AuthorizableTest::testRemoveGroupNoGroups                            | 00:00:00.94   | 00:00:00.50 |
+--------------------------------------------------------------------------------------------+---------------+-------------+
...and 108 more tests hidden from view.


Time: 02:02.688, Memory: 42.00 MB

OK (225 tests, 513 assertions)
```

**Review by MGatner — APPROVED** [review]

**Pull request merged**